File: ambeth/Ambeth.Cache.Bytecode/ambeth/bytecode/behavior/CacheMapEntryBehavior.cs
using De.Osthus.Ambeth.Bytecode.Visitor;
using De.Osthus.Ambeth.Cache.Collections;
using De.Osthus.Ambeth.Ioc.Annotation;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Merge;
using De.Osthus.Ambeth.Merge.Model;
using System.Collections.Generic;

namespace De.Osthus.Ambeth.Bytecode.Behavior
{
    public class CacheMapEntryBehavior : AbstractBehavior
    {
        [LogInstance]
        public ILogger Log { private get; set; }

        [Autowired]
        public IEntityMetaDataProvider EntityMetaDataProvider { protected get; set; }

        public override IClassVisitor Extend(IClassVisitor visitor, IBytecodeBehaviorState state, IList<IBytecodeBehavior> remainingPendingBehaviors,
                    IList<IBytecodeBehavior> cascadePendingBehaviors)
        {
            CacheMapEntryEnhancementHint hint = state.GetContext<CacheMapEntryEnhancementHint>();
            if (hint == null)
            {
                return visitor;
            }
            IEntityMetaData metaData = EntityMetaDataProvider.GetMetaData(hint.GetEntityType());
            visitor = new CacheMapEntryVisitor(visitor, metaData, hint.GetIdIndex());
            return visitor;
        }
    }
}

File: jambeth/jAmbeth.Cache.Bytecode/src/de/osthus/ambeth/bytecode/behavior/CacheMapEntryBehavior.java
package de.osthus.ambeth.bytecode.behavior;

import java.util.List;

import de.osthus.ambeth.bytecode.visitor.CacheMapEntryVisitor;
import de.osthus.ambeth.cache.collections.CacheMapEntryEnhancementHint;
import de.osthus.ambeth.ioc.annotation.Autowired;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.merge.IEntityMetaDataProvider;
import de.osthus.ambeth.merge.model.IEntityMetaData;
import de.osthus.ambeth.repackaged.org.objectweb.asm.ClassVisitor;

public class CacheMapEntryBehavior extends AbstractBehavior
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	@Autowired
	protected IEntityMetaDataProvider entityMetaDataProvider;

	@Override
	public ClassVisitor extend(ClassVisitor visitor, IBytecodeBehaviorState state, List<IBytecodeBehavior> remainingPendingBehaviors,
			List<IBytecodeBehavior> cascadePendingBehaviors)
	{
		final CacheMapEntryEnhancementHint hint = state.getContext(CacheMapEntryEnhancementHint.class);
		if (hint == null)
		{
			return visitor;
		}
		IEntityMetaData metaData = entityMetaDataProvider.getMetaData(hint.getEntityType());
		visitor = new CacheMapEntryVisitor(visitor, metaData, hint.getIdIndex());
		return visitor;
	}
}
